File: Litvinov group/semester 3/Lazy/Lazy/LazyFactory.cs
using System;

namespace Lazy
{
    /// <summary>
    /// Class that creates Lazy classes based on function supplier
    /// </summary>
    public class LazyFactory
    {
        /// <summary>
        /// Create Lazy class without multithreading
        /// </summary>
        /// <typeparam name="T">Answer type</typeparam>
        /// <param name="supplier">Calculations function</param>
        /// <returns>Result of the calculations</returns>
        public static Lazy<T> CreateLazy<T>(Func<T> supplier) => new Lazy<T>(supplier);

        /// <summary>
        /// Create MultiThreadLazy class with semithreading
        /// </summary>
        /// <typeparam name="T">Answer type</typeparam>
        /// <param name="supplier">Calculations function</param>
        /// <returns>Result of the calculations</returns>
        public static MultiThreadLazy<T> CreateMultiThreadLazy<T>(Func<T> supplier) 
            => new MultiThreadLazy<T>(supplier);
    }
}

File: Litvinov group/semester 4/08.HW/08.HW/Lazy/LazyFactory.fs
module LazyFactory

open ILazy
open SingleThreadLazy
open MultiThreadLazy
open MultiThreadLazyLockFree

/// Class that creates Lazy classes based on function supplier
type LazyFactory =

    /// Create Lazy class without multithreading
    static member CreateSingleThreadLazy supplier =
        new SingleThreadLazy<'a>(supplier) :> ILazy<'a>

    /// Create MultiThreadLazy class with semithreading
    static member CreateMultiThreadLazy supplier = 
        new MultiThreadLazy<'a>(supplier) :> ILazy<'a>

    /// Create MultiThreadLazy class with semithreading without locks
    static member CreateMultiThreadLazyLockFree supplier =
        new MultiThreadLazyLockFree<'a>(supplier) :> ILazy<'a>
